addOns/paramminer/src/main/java/org/zaproxy/addon/paramminer/GuesserScan.java

```diff
@@ -0,0 +1,264 @@
+ * limitations under the License.
+ */
+package org.zaproxy.addon.paramminer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.swing.table.TableModel;
+import org.zaproxy.addon.paramminer.gui.ParamMinerHistoryTableModel;
+import org.zaproxy.zap.model.GenericScanner2;
+
+public class GuesserScan implements GenericScanner2 {
+    private static enum State {
+        NOT_STARTED,
+        RUNNING,
+        PAUSED,
+        STOPPED,
+        FINISHED
+    }
+
+    private final Lock lock;
+    private int scanId;
+    private State state;
+
+    private String displayName;
+    private ParamMinerConfig config;
+
+    private int tasksDoneCount;
+    private int tasksTodoCount;
+
+    private final ExecutorService threadPool;
+    private final List<GuesserProgressListener> listeners;
+    private final ParamMinerHistoryTableModel tableModel;
+    private String output;
+
+    public GuesserScan(int scanId, ParamMinerConfig config, String name) {
+        this.scanId = scanId;
+        this.displayName = name;
+        this.lock = new ReentrantLock();
+        this.state = State.NOT_STARTED;
+        this.config = config;
+
+        listeners = new ArrayList<>(2);
+        tableModel = new ParamMinerHistoryTableModel();
+        output = "scanId: " + scanId;
+
+        this.threadPool =
+                Executors.newFixedThreadPool(
+                        config.getThreadpoolSize(),
+                        new ParamGuesserThreadFactory("ZAP-ParamGuesser-" + scanId + "-thread-"));
+    }
+
+    public void addProgressListener(GuesserProgressListener listener) {
+        listeners.add(listener);
+    }
+
+    public TableModel getTableModel() {
+        return tableModel;
+    }
+
+    public String getOutput() {
+        return output;
+    }
+
+    public ParamMinerConfig getConfig() {
+        return config;
+    }
+
+    @Override
+    public void run() {}
+
+    @Override
+    public void setScanId(int id) {
+        this.scanId = id;
+    }
+
+    @Override
+    public int getScanId() {
+        return scanId;
+    }
+
+    public String getState() {
+        lock.lock();
+        try {
+            return state.toString();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void setDisplayName(String name) {
+        this.displayName = name;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return this.displayName;
+    }
+
+    public void start() {
+        lock.lock();
+        try {
+            if (state == State.NOT_STARTED) {
+                state = State.RUNNING;
+
+                // TODO start
+            }
+        } finally {
+            lock.unlock();
+        }
```

Comment:
Should state be set here? I mean if it can't be started for some reason should it go back to NOT_STARTED or FINISHED?